src/avram/errors.cr

```diff
@@ -14,40 +14,53 @@ module Avram
   # Raised when trying to access a record that was not preloaded and lazy load
   # is disabled.
   class LazyLoadError < AvramError
-    def initialize(model : String, association : String)
+    getter model
+    getter association
+
+    def initialize(@model : String, @association : String)
       super "#{association} for #{model} must be preloaded with 'preload_#{association}'"
     end
   end
 
   # Raised when Avram cannot find a record by given id
   class RecordNotFoundError < AvramError
-    def initialize(model : TableName, id : String)
+    getter model
+    getter id = nil
+    getter query = nil
+
+    def initialize(@model : TableName, @id : String)
       super "Could not find #{model} with id of #{id}"
     end
 
-    def initialize(model : TableName, query : Symbol)
+    def initialize(@model : TableName, @query : Symbol)
       super "Could not find #{query} record in #{model}"
     end
   end
 
   class MissingRequiredAssociationError < AvramError
-    def initialize(model : Avram::Model.class, association : Avram::Model.class)
+    getter model
+    getter association
+
+    def initialize(@model : Avram::Model.class, @association : Avram::Model.class)
       super "Expected #{model} to have an association with #{association} but one was not found."
     end
   end
 
   # Raised when a validation is expecting an impossible constraint
   class ImpossibleValidation < AvramError
-    def initialize(attribute : Symbol, message = "an impossible validation")
+    getter attribute
+
+    def initialize(@attribute : Symbol, message = "an impossible validation")
       super "Validation for #{attribute} can never satisfy #{message}"
     end
   end
 
   # Raised when using the create!, update!, or delete! methods on an operation when it does not have the proper attributes
-  class InvalidOperationError < AvramError
+  class InvalidOperationError(T) < AvramError
+    getter operation
     getter errors : Hash(Symbol, Array(String))
 
-    def initialize(operation)
+    def initialize(@operation : T)
```

Comment:
It did not compiler. I'm not sure, isn't that because it is just a generic method, and the compiler will not allow me to add a property for `operation` to the class? I also tried adding a property of type `Reference`, but the compiler insists that I choose a more specific type...

Reply:
Oh, yeah, that makes sense. I was thinking this might be a breaking change, but I just tried it out and it didn't seem to be. I think we can leave it as you have it.